File: solutions/cpp/Levenshtein_distance.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <numeric>
#include <random>
#include <string>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::min;
using std::random_device;
using std::string;
using std::uniform_int_distribution;
using std::vector;

// @include
int LevenshteinDistance(string A, string B) {
  // Try to reduce the space usage.
  if (A.size() < B.size()) {
    swap(A, B);
  }

  vector<int> D(B.size() + 1);
  // Initialization.
  iota(D.begin(), D.end(), 0);

  for (int i = 1; i <= A.size(); ++i) {
    int pre_i_1_j_1 = D[0];  // Stores the value of D[i - 1][j - 1].
    D[0] = i;
    for (int j = 1; j <= B.size(); ++j) {
      int pre_i_1_j = D[j];  // Stores the value of D[i -1][j].
      D[j] = A[i - 1] == B[j - 1] ? pre_i_1_j_1
                                  : 1 + min(pre_i_1_j_1, min(D[j - 1], D[j]));
      // Previous D[i - 1][j] will become the next D[i - 1][j - 1].
      pre_i_1_j_1 = pre_i_1_j;
    }
  }
  return D.back();
}
// @exclude

string RandString(int len) {
  default_random_engine gen((random_device())());
  uniform_int_distribution<int> dis('a', 'z');
  string ret;
  while (len--) {
    ret += dis(gen);
  }
  return ret;
}

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  string A, B;
  // Wiki example (http://en.wikipedia.org/wiki/Levenshtein_distance)
  A = "Saturday", B = "Sunday";
  assert(3 == LevenshteinDistance(A, B));
  A = "kitten", B = "sitting";
  assert(3 == LevenshteinDistance(A, B));

  if (argc == 3) {
    A = argv[1], B = argv[2];
  } else {
    uniform_int_distribution<int> dis(1, 100);
    A = RandString(dis(gen));
    B = RandString(dis(gen));
  }
  cout << A << endl << B << endl;
  cout << LevenshteinDistance(A, B) << endl;
  return 0;
}

File: solutions/python/levenshtein_distance.py
import sys
import random
import string


# @include
def levenshtein_distance(A, B):
    def compute_distance_between_prefixes(A_idx, B_idx):
        if A_idx < 0:
            # A is empty so add all of B's characters.
            return B_idx + 1
        elif B_idx < 0:
            # B is empty so delete all of A's characters.
            return A_idx + 1
        if distance_between_prefixes[A_idx][B_idx] == -1:
            if A[A_idx] == B[B_idx]:
                distance_between_prefixes[A_idx][B_idx] = (
                    compute_distance_between_prefixes(A_idx - 1, B_idx - 1))
            else:
                substitute_last = compute_distance_between_prefixes(A_idx - 1,
                                                                    B_idx - 1)
                add_last = compute_distance_between_prefixes(A_idx - 1, B_idx)
                delete_last = compute_distance_between_prefixes(A_idx,
                                                                B_idx - 1)
                distance_between_prefixes[A_idx][B_idx] = (
                    1 + min(substitute_last, add_last, delete_last))
        return distance_between_prefixes[A_idx][B_idx]

    distance_between_prefixes = [[-1] * len(B) for _ in A]
    return compute_distance_between_prefixes(len(A) - 1, len(B) - 1)
# @exclude


def check_answer(A, B):
    # Try to reduce the space usage.
    if len(A) < len(B):
        A, B = B, A

    # Initialization.
    D = list(range(len(B) + 1))
    for i in range(1, len(A) + 1):
        pre_i_1_j_1 = D[0]  # Stores the value of D[i - 1][j - 1].
        D[0] = i
        for j in range(1, len(B) + 1):
            pre_i_1_j = D[j]  # Stores the value of D[i -1][j].
            D[j] = pre_i_1_j_1 if A[i - 1] == B[j - 1] else 1 + min(
                pre_i_1_j_1, D[j - 1], D[j])
            # Previous D[i - 1][j] will become the next D[i - 1][j - 1].
            pre_i_1_j_1 = pre_i_1_j
    return D[-1]


def rand_string(length):
    return ''.join(random.choice(string.ascii_lowercase) for _ in range(length))


def main():
    # Wiki example (http:#en.wikipedia.org/wiki/levenshtein_distance)
    A, B = 'k', 'sitt'
    assert 4 == levenshtein_distance(A, B)
    A, B = 'Saturday', 'Sunday'
    assert 3 == levenshtein_distance(A, B)
    A, B = 'kitten', 'sitting'
    assert 3 == levenshtein_distance(A, B)

    if len(sys.argv) == 3:
        A, B = sys.argv[1:]
    else:
        A = rand_string(random.randint(1, 100))
        B = rand_string(random.randint(1, 100))
    print(A)
    print(B)
    print(levenshtein_distance(A, B))
    assert levenshtein_distance(A, B) == check_answer(A, B)


if __name__ == '__main__':
    main()
